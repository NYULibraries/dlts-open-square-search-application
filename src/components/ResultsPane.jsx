import React from "react";
import PropTypes from "prop-types";
import { ResultItem } from "./ResultItem";

// container component that renders multiple ResultItem components
// or displays helper texts for errors and empty searches
export function ResultsPane({
    publications = [],
    error = false,
    numBooks = 0,
}) {
    console.log("numBooks ", numBooks);
    const listItems = publications.map((book, i) => (
        <ResultItem
            contributors={book.contributors}
            date={book.date}
            description={book.description}
            identifier={book.id}
            index={book.id + i}
            key={book.id + i}
            maxDescriptionLength={500}
            subtitle={book.subtitle}
            title={book.title}
        />
    ));
    return (
        <section>
            <header>
                <h2 className="osq-resultsheader">
                    {error &&
                        "Sorry, a server error has occurred. Please try your search again later.\n"}
                    {publications.length > 0
                        ? `Results: ${publications.length} books`
                        : "Results: None"}
                </h2>
            </header>
            <div>
                {publications.length > 0 ? (
                    listItems
                ) : (
                    <span>Please try another search.</span>
                )}
            </div>
        </section>
    );
}

ResultsPane.propTypes = {
    // books: PropTypes.array,
    // they are not books, they are publications of different kinds of media
    publications: PropTypes.array,
    // use these
    // NOTE: this component in react doesn't even need to know if it should display
    // rendering logic should be carried in parent component
    // display: PropTypes.Boolean.required,
    // Ellipsis character -- on Macos use key combination `Option + ;`
    // ellipsis: PropTypes.String,
    error: PropTypes.bool,
    // explore this shape a little more
    // TODO: does this belong on the pane or does it belong on the item
    // highlights: PropTypes.Object.required,
    // TODO: does this belong on the pane or does it belong on the item
    // maxDescriptionLength: PropTypes.number,
    numBooks: PropTypes.number.isRequired,
    // results: PropTypes.array.required,
};
