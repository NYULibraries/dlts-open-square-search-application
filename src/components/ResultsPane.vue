<template>
    <div class="osq-results-hold">
        <!--RESULTS-->
        <template v-show="display">
            <!-- v-show is necessary on this <header> element for some reason.
                 Its visibility is not being toggled along with the other elements
                 in this <template>.
            -->
            <header v-show="display">
                <h2 class="osq-resultsheader">
                    {{ resultsHeader }}
                </h2>
            </header>
            <div
                v-show="display"
                class="osq-results"
            >
                <span v-if="( ! error ) && ( ! results || results.length === 0 )">
                    Please try another search.
                </span>

                <!--BOOK-->
                <div
                    v-for="result in results"
                    :id="result.identifier"
                    :key="result.identifier"
                    :name="result.title"
                >
                    <article class="book-summary">
                        <div
                            class="thumb"
                            role="presentation"
                        >
                            <a
                                :href="'/books/' + result.identifier + '/'"
                                tabindex="-1"
                                aria-hidden="true"
                            >
                                <img
                                    :src="`/open-square-reader/epub_content/${ result.identifier }/ops/images/${ result.identifier }-th.jpg`"
                                    alt=""
                                    class="img-fluid"
                                >
                            </a>
                        </div>
                        <div class="meta">
                            <div class="book-title-group">
                                <a
                                    :href="`/books/${ result.identifier }`"
                                >
                                    <div
                                        class="book-title"
                                        v-html="getFieldValueOrHighlightedFieldValue( result, 'title' )"
                                    >
                                    </div>
                                    <div
                                        class="book-subtitle"
                                        v-html="getFieldValueOrHighlightedFieldValue( result, 'subtitle' )"
                                    >
                                    </div>
                                </a>
                            </div>

                            <div
                                class="author"
                                v-html="getFieldValueOrHighlightedFieldValue( result, 'author' )"
                            >
                            </div>
                            <div class="pubdate">
                                <span>Published:</span> <span v-html="getFieldValueOrHighlightedFieldValue( result, 'date' )"></span>
                            </div>
                            <div
                                class="meta"
                                v-html="getFieldValueOrHighlightedFieldValue( result, 'description' )"
                            >
                            </div>
                        </div>
                    </article>
                </div>
                <!--BOOK-->
            </div>
        </template>
        <!--RESULTS-->
    </div>
</template>

<script>
export default {
    name   : 'ResultsPane',
    props  : {
        display  : {
            type     : Boolean,
            required : true,
            default  : false,
        },
        error    : {
            type     : Boolean,
            required : true,
            default  : false,
        },
        highlights   : {
            type     : Object,
            required : true,
            default  : function () {
                return null;
            },
        },
        numBooks : {
            type     : Number,
            required : true,
            default  : null,
        },
        results  : {
            type     : Array,
            required : true,
            default  : function () {
                return null;
            },
        },
    },
    data() {
        return {
            publicPath : process.env.BASE_URL,
        };
    },
    computed : {
        numBooksFormatted : function () {
            return this.numBooks ? this.numBooks.toLocaleString() : '';
        },
        resultsHeader : function () {
            if ( this.error ) {
                return 'Sorry, a server error has occurred.  Please try your search again later.';
            }

            if ( this.results && this.results.length > 0 ) {
                return 'Results: ' + this.numBooksFormatted + ' books';
            } else {
                return 'Results: None';
            }
        },
    },
    methods : {
        getFieldValueOrHighlightedFieldValue( result, field ) {
            const identifier = result.identifier;

            if ( this.highlights[ identifier ] && this.highlights[ identifier ][ field ] ) {
                // We only want the first snippet
                return this.highlights[ identifier ][ field ][ 0 ];
            } else {
                const fieldValue = result[ field ];

                if ( Array.isArray( fieldValue ) ) {
                    return fieldValue[ 0 ];
                } else {
                    return fieldValue;
                }
            }
        },
    },
};
</script>
