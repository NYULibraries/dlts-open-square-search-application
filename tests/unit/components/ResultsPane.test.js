import ResultsPane from '@/components/ResultsPane';

import merge from 'lodash.merge';
import { shallowMount } from '@vue/test-utils';

function createWrapper( overrides ) {
    const defaultMountingOptions = {
        propsData : {
            display              : false,
            error                : false,
            highlights           : {},
            numBooks             : 0,
            results              : [],
        },
    };

    return shallowMount( ResultsPane, merge( defaultMountingOptions, overrides ) );
}

describe( 'ResultsPane', () => {
    const RESULTS =
        require( '../fixtures/solr-responses/solr-search.json' );

    let wrapper;

    function simulateError() {
        wrapper.setProps(
            {
                display : true,
                error   : true,
            }
        );
    }

    function simulateSearch() {
        wrapper.setProps(
            {
                display    : true,
                highlights : RESULTS.highlighting,
                numBooks   : RESULTS.length,
                results    : RESULTS.response.docs,
            }
        );
    }

    beforeEach( () => {
        wrapper = createWrapper();
    } );

    test( 'renders no results found correctly', () => {
        wrapper.setProps(
            {
                display  : true,
                numBooks : 0,
                results  : [],
            }
        );

        expect( wrapper.element ).toMatchSnapshot();
    } );

    test( 'renders results correctly', () => {
        simulateSearch();

        expect( wrapper.element ).toMatchSnapshot();
    } );

    test( 'renders error condition correctly', () => {
        simulateError();

        expect( wrapper.element ).toMatchSnapshot();
    } );
} );
