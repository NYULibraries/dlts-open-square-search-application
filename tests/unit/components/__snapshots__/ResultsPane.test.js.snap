// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsPane renders error condition correctly 1`] = `
<div
  class="osq-results-hold"
>
  <header
    style=""
  >
    <h2
      class="osq-resultsheader"
    >
      
                Sorry, a server error has occurred.  Please try your search again later.
            
    </h2>
  </header>
   
  <div
    class="osq-results"
    style=""
  >
    <!---->
     
  </div>
</div>
`;

exports[`ResultsPane renders no results found correctly 1`] = `
<div
  class="osq-results-hold"
>
  <header
    style=""
  >
    <h2
      class="osq-resultsheader"
    >
      
                Results: None
            
    </h2>
  </header>
   
  <div
    class="osq-results"
    style=""
  >
    <span>
      
                Please try another search.
            
    </span>
     
  </div>
</div>
`;

exports[`ResultsPane renders results correctly 1`] = `
<div
  class="osq-results-hold"
>
  <header
    style=""
  >
    <h2
      class="osq-resultsheader"
    >
      
                Results:  books
            
    </h2>
  </header>
   
  <div
    class="osq-results"
    style=""
  >
    <!---->
     
    <div
      class="book-summary-hold"
      id="9781111111111"
      name="Title 1 Highlighting"
    >
      <article
        class="book-summary"
      >
        <div
          class="thumb"
          role="presentation"
        >
          <a
            aria-hidden="true"
            href="/books/9781111111111/"
            tabindex="-1"
          >
            <img
              alt=""
              class="img-fluid"
              src="/open-square-reader/epub_content/9781111111111/ops/images/9781111111111-th.jpg"
            />
          </a>
        </div>
         
        <div
          class="meta"
        >
          <div
            class="book-title-group"
          >
            <a
              href="/books/9781111111111"
            >
              <div
                class="book-title"
              >
                Title 1 
                <mark>
                  Highlighting
                </mark>
              </div>
               
              <div
                class="book-subtitle"
              >
                Subtitle 1
              </div>
            </a>
          </div>
           
          <div
            class="author"
          >
            Author One with Author Two, Author Three, Author Four, Author Five and Author Six
          </div>
           
          <div
            class="pubdate"
          >
            <span>
              Published: 
            </span>
            <span>
              2018
            </span>
          </div>
           
          <div
            class="description meta"
          >
            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Duis ut diam quam nulla porttitor. Tellus in hac habitasse platea dictumst vestibulum rhoncus est. Pharetra massa massa ultricies mi quis hendrerit dolor magna eget. Tincidunt dui ut ornare lectus sit amet est. Quam quisque id diam vel quam elementum pulvinar etiam non. Quisque non tellus orci ac auctor augue mauris augue neque. Mi ipsum faucibus vitae aliquet nec…
          </div>
        </div>
      </article>
    </div>
    <div
      class="book-summary-hold"
      id="9781111111112"
      name="Title 2"
    >
      <article
        class="book-summary"
      >
        <div
          class="thumb"
          role="presentation"
        >
          <a
            aria-hidden="true"
            href="/books/9781111111112/"
            tabindex="-1"
          >
            <img
              alt=""
              class="img-fluid"
              src="/open-square-reader/epub_content/9781111111112/ops/images/9781111111112-th.jpg"
            />
          </a>
        </div>
         
        <div
          class="meta"
        >
          <div
            class="book-title-group"
          >
            <a
              href="/books/9781111111112"
            >
              <div
                class="book-title"
              >
                Title 2
              </div>
               
              <div
                class="book-subtitle"
              >
                Subtitle 2 
                <mark>
                  Highlights
                </mark>
              </div>
            </a>
          </div>
           
          <div
            class="author"
          >
            Author Seven, Author Eight
          </div>
           
          <div
            class="pubdate"
          >
            <span>
              Published: 
            </span>
            <span>
              2016
            </span>
          </div>
           
          <div
            class="description meta"
          >
            Rhoncus est pellentesque elit ullamcorper dignissim cras. Potenti nullam ac tortor vitae purus faucibus. Pulvinar mattis nunc sed blandit libero volutpat sed cras. Aliquam sem et tortor consequat. Est ante in nibh mauris cursus mattis molestie a. Risus in hendrerit gravida rutrum quisque non tellus orci ac. Arcu non sodales neque sodales ut etiam. Posuere lorem ipsum dolor sit amet. Adipiscing elit ut aliquam purus sit amet luctus venenatis lectus. Tempor id eu nisl nunc. Risus at ultrices mi…
          </div>
        </div>
      </article>
    </div>
    <div
      class="book-summary-hold"
      id="9781111111113"
      name="Title 3"
    >
      <article
        class="book-summary"
      >
        <div
          class="thumb"
          role="presentation"
        >
          <a
            aria-hidden="true"
            href="/books/9781111111113/"
            tabindex="-1"
          >
            <img
              alt=""
              class="img-fluid"
              src="/open-square-reader/epub_content/9781111111113/ops/images/9781111111113-th.jpg"
            />
          </a>
        </div>
         
        <div
          class="meta"
        >
          <div
            class="book-title-group"
          >
            <a
              href="/books/9781111111113"
            >
              <div
                class="book-title"
              >
                Title 3
              </div>
               
              <div
                class="book-subtitle"
              >
                Subtitle 3
              </div>
            </a>
          </div>
           
          <div
            class="author"
          >
            Author 
            <mark>
              Highlight
            </mark>
            , Author Nine
          </div>
           
          <div
            class="pubdate"
          >
            <span>
              Published: 
            </span>
            <span>
              2019
            </span>
          </div>
           
          <div
            class="description meta"
          >
            Scelerisque felis imperdiet proin fermentum. Netus et malesuada fames ac turpis. Phasellus faucibus scelerisque eleifend donec. Velit scelerisque in dictum non consectetur a erat nam at. Ut sem nulla pharetra diam sit amet nisl suscipit. Velit ut tortor pretium viverra suspendisse potenti. Mi in nulla posuere sollicitudin. Fringilla phasellus faucibus scelerisque eleifend donec pretium vulputate sapien. In eu mi bibendum neque. Id consectetur purus ut faucibus pulvinar elementum integer.…
          </div>
        </div>
      </article>
    </div>
    <div
      class="book-summary-hold"
      id="9781111111114"
      name="Title 4"
    >
      <article
        class="book-summary"
      >
        <div
          class="thumb"
          role="presentation"
        >
          <a
            aria-hidden="true"
            href="/books/9781111111114/"
            tabindex="-1"
          >
            <img
              alt=""
              class="img-fluid"
              src="/open-square-reader/epub_content/9781111111114/ops/images/9781111111114-th.jpg"
            />
          </a>
        </div>
         
        <div
          class="meta"
        >
          <div
            class="book-title-group"
          >
            <a
              href="/books/9781111111114"
            >
              <div
                class="book-title"
              >
                Title 4
              </div>
               
              <div
                class="book-subtitle"
              >
                Subtitle 4
              </div>
            </a>
          </div>
           
          <div
            class="author"
          >
            Author Ten, Author Eleven, Author Twelve, Author Thirteen, Author Fourteen and Author Fifteen
          </div>
           
          <div
            class="pubdate"
          >
            <span>
              Published: 
            </span>
            <span>
              <mark>
                2000
              </mark>
            </span>
          </div>
           
          <div
            class="description meta"
          >
            Egestas fringilla phasellus faucibus scelerisque eleifend donec pretium. Aliquam ut porttitor leo a diam sollicitudin. In dictum non consectetur a. Fringilla ut morbi tincidunt augue interdum velit. Risus at ultrices mi tempus imperdiet nulla. Risus quis varius quam quisque id. Natoque penatibus et magnis dis parturient montes nascetur. Nibh praesent tristique magna sit amet purus gravida. Eget lorem dolor sed viverra ipsum nunc aliquet bibendum. Egestas integer eget aliquet nibh praesent…
          </div>
        </div>
      </article>
    </div>
    <div
      class="book-summary-hold"
      id="9781111111115"
      name="Title 5"
    >
      <article
        class="book-summary"
      >
        <div
          class="thumb"
          role="presentation"
        >
          <a
            aria-hidden="true"
            href="/books/9781111111115/"
            tabindex="-1"
          >
            <img
              alt=""
              class="img-fluid"
              src="/open-square-reader/epub_content/9781111111115/ops/images/9781111111115-th.jpg"
            />
          </a>
        </div>
         
        <div
          class="meta"
        >
          <div
            class="book-title-group"
          >
            <a
              href="/books/9781111111115"
            >
              <div
                class="book-title"
              >
                Title 5
              </div>
               
              <div
                class="book-subtitle"
              >
                Subtitle 5
              </div>
            </a>
          </div>
           
          <div
            class="author"
          >
            Author Sixteen, Author Seventeen, Author Eighteen, Author Nineteen, Author Twenty
          </div>
           
          <div
            class="pubdate"
          >
            <span>
              Published: 
            </span>
            <span>
              2016
            </span>
          </div>
           
          <div
            class="description meta"
          >
            …In dictum 
            <mark>
              HIGHLIGHT
            </mark>
             non consectetur a erat nam at lectus urna. Euismod in pellentesque massa 
            <mark>
              highlights
            </mark>
             placerat duis ultricies lacus. Auctor urna nunc id cursus metus aliquam eleifend 
            <mark>
              Highlight
            </mark>
             mi in. Sapien pellentesque habitant morbi tristique senectus et netus et. Sed enim ut sem viverra aliquet. Tempor id eu nisl nunc mi ipsum faucibus vitae aliquet. Interdum velit laoreet id donec ultrices tincidunt arcu non. Sociis natoque penatibus et magnis dis parturient montes. Vulputate ut pharetra sit amet aliquam id diam maecenas. Elit duis tristique sollicitudin nibh sit amet commodo nulla.…
          </div>
        </div>
      </article>
    </div>
  </div>
</div>
`;
